docs/manual/src/udl/callback_interfaces.md

```diff
@@ -1,107 +1,55 @@
-  - Storing `Box<dyn CallbackInterfaceTrait>` types inside a type that needs to be `Send + Sync` (for example a UniFFI
-    interface type)
-  - Storing `Box<dyn CallbackInterfaceTrait>` inside a global `Mutex` or `RwLock`
-
-
-## 2. Setup error handling
-
-All methods of the Rust trait should return a Result.  The error half of that result must
-be an [error type defined in the UDL](./errors.md).
-
-It's currently allowed for callback interface methods to return a regular value
-rather than a `Result<>`.  However, this is means that any exception from the
-foreign bindings will lead to a panic.
-
-### Extra requirements for errors used in callback interfaces
 
-In order to support errors in callback interfaces, UniFFI must be able to
-properly [lift the error](../internals/lifting_and_lowering.md).  This means
-that the if the error is described by an `enum` rather than an `interface` in
-the UDL (see [Errors](./errors.md)) then all variants of the Rust enum must be unit variants.
-
-In addition to expected errors, a callback interface call can result in all kinds of
-unexpected errors.  Some examples are the foreign code throws an exception that's not part
-of the exception type or there was a problem marshalling the data for the call.  UniFFI
-uses `uniffi::UnexpectedUniFFICallbackError` for these cases.  Your code must include a
-`From<uniffi::UnexpectedUniFFICallbackError>` impl for your error type to handle those or
-the UniFFI scaffolding code will fail to compile.  See `example/callbacks` for an
-example of how to do this.
-
-## 3. Define a callback interface in the UDL
-
-```webidl
-callback interface Keychain {
-    [Throws=KeyChainError]
-    string? get(string key);
-
-    [Throws=KeyChainError]
-    void put(string key, string data);
-};
+  // Async API
+  async fn get_async(&self, key: String) -> Result<Option<String>, KeyChainError>
+  async fn put_async(&self, key: String, value: String) -> Result<(), KeyChainError>
+}
 ```
 
-## 4. And allow it to be passed into Rust
+Note: Rust doesn't support async functions in traits yet.  The proc-macro
+converts async function signatures into a sync function that returns
+`uniffi::ForeignFuture<T>`, similar to the `async-trait` crate.
+`uniffi::ForeignFuture<T>` implements `Future<Output = T>` and `Send`.
 
-Here, we define a constructor to pass the keychain to rust, and then another method
-which may use it.
+## Setup error handling
 
-In UDL:
-
-```webidl
-interface Authenticator {
-    constructor(Keychain keychain);
-    void login();
-}
-```
+All methods of the Rust trait should return a Result.  This is needed because
+any callback method call can fail unexpected ways that shouldn't result in a
+panic, for example when the foreign code throws an exception we weren't
+expecting.
 
-In Rust:
+If your method can only fail in these unexpected ways, use
+`uniffi::CallbackResult<T>` as your result type.  This is a type alias for
+`Result<T, uniffi::UnexpectedCallbackError>`.
```

Comment:
This is a small change that I think could improve ergonomics.  If a callback interface method isn't expected to raise an error, then we shouldn't need to define an error type for it.